src/txrequest.cpp

```diff
@@ -0,0 +1,715 @@
+            // Can have at most 1 CANDIDATE_BEST/REQUESTED peer
+            assert(c.m_candidate_best + c.m_requested <= 1);
+            // If there are any CANDIDATE_READY entries, there must be exactly one CANDIDATE_BEST or REQUESTED
+            // entry.
+            if (c.m_candidate_ready > 0) {
+                assert(c.m_candidate_best + c.m_requested == 1);
+            }
+            // If there is both a CANDIDATE_READY and a CANDIDATE_BEST entry, the CANDIDATE_BEST one must be at
+            // least as good as the best CANDIDATE_READY.
+            if (c.m_candidate_ready && c.m_candidate_best) {
+                assert(c.m_priority_candidate_best <= c.m_priority_best_candidate_ready);
+            }
+            // Detect duplicate (peer, txid) entries
+            std::sort(c.m_peers.begin(), c.m_peers.end());
+            assert(std::adjacent_find(c.m_peers.begin(), c.m_peers.end()) == c.m_peers.end());
+            // Verify all per_txhash flags.
+            uint8_t expected_per_txhash = 0;
+            if (c.m_any_preferred_first || c.m_requested) {
+                expected_per_txhash |= TXHASHINFO_NO_MORE_PREFERRED_FIRST;
+            }
+            if (c.m_any_nonpreferred_first || c.m_requested) {
+                expected_per_txhash |= TXHASHINFO_NO_MORE_NONPREFERRED_FIRST;
+            }
+            // All expected flags must be present, but there can be more. If a node went from REQUESTED to
+            // COMPLETED, or was deleted, our expected_per_txhash may miss the relevant bits.
+            assert((expected_per_txhash & ~c.m_or_all_per_txhash) == 0);
+            // No entry can have flags that are a superset of the actual ones (they're always ORed into the actual
+            // one).
+            auto it_last = std::prev(m_index.get<ByTxHash>().lower_bound(
+                EntryTxHash{entry.first, State::TOO_LARGE, 0}));
+            assert(it_last->m_txhash == entry.first);
+            assert(c.m_or_all_per_txhash == it_last->m_per_txhash);
+        }
+    }
+
+    void PostGetRequestableSanityCheck(std::chrono::microseconds now) const
+    {
+        for (const auto& entry : m_index) {
+            if (entry.IsWaiting()) {
+                // REQUESTED and CANDIDATE_DELAYED must have a time in the future (they should have been converted
+                // to COMPLETED/CANDIDATE_READY respectively).
+                assert(entry.m_time > now);
+            } else if (entry.IsSelectable()) {
+                // CANDIDATE_READY and CANDIDATE_BEST cannot have a time in the future (they should have remained
+                // CANDIDATE_DELAYED, or should have been converted back to it if time went backwards).
+                assert(entry.m_time <= now);
+            }
+        }
+    }
+};
+
+TxRequestTracker::TxRequestTracker(bool deterministic)
+{
+    m_impl = MakeUnique<TxRequestTracker::Impl>(deterministic);
```

Comment:
Consider using an initializer list here instead of setting `m_impl` in the ctor function body.

If you did that, you could make `m_impl` const, which is fine since you've deleted the move constructor and assignment.

`const unique_ptr` communicates that _this_ `m_impl` is always the object that `TxRequestTracker` owns. See Herb Sutter's Leak-Freedom talk at https://youtu.be/JfmTagWcqoE?t=571

```diff
diff --git a/src/txrequest.cpp b/src/txrequest.cpp
index f6c387f8f4..b7d90a1ba5 100644
--- a/src/txrequest.cpp
+++ b/src/txrequest.cpp
@@ -668,10 +668,8 @@ public:
     }
 };
 
-TxRequestTracker::TxRequestTracker(bool deterministic)
-{
-    m_impl = MakeUnique<TxRequestTracker::Impl>(deterministic);
-}
+TxRequestTracker::TxRequestTracker(bool deterministic) :
+    m_impl{MakeUnique<TxRequestTracker::Impl>(deterministic)} {}
 
 TxRequestTracker::~TxRequestTracker() {}
 
diff --git a/src/txrequest.h b/src/txrequest.h
index 6e892eebe1..83f6dbe2fe 100644
--- a/src/txrequest.h
+++ b/src/txrequest.h
@@ -119,7 +119,7 @@ public:
 private:
     // Avoid littering this header file with implementation details.
     class Impl;
-    std::unique_ptr<Impl> m_impl;
+    const std::unique_ptr<Impl> m_impl;
 
 public:
     //! Construct a TxRequestTracker.
```

Reply:
Done.